### PR TITLE
improve GPU support

### PR DESCRIPTION
**Pull request by mtar**
## Description

This fix enables the usage of GPU-tensors in heat if ht.use_device is globally set to "gpu". The existing unit tests are modified to test the cpu or gpu only. Also, tensors on locally set devices can be tested.

Fixes: #

Changes proposed:
– unit tests depend on selected device (Environment variable DEVICE=gpu for global, lgpu for locally defined gpu tensors, cpu otherwise)
– additional copy operations GPU <-> CPU for not cuda-aware MPI
– smaller fixes on some core functions for accepting gpu tensors

## Type of change
• Setting device parameter in torch functions
• bugfixes
• test modifications

Select relevant options.

[x] Bug fix (non-breaking change which fixes an issue)
[ ] New feature (non-breaking change which adds functionality)
[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
[ ] Documentation update

Are all split configurations tested and accounted for?
[x] yes [ ] no
Does this change require a documentation update outside of the changes proposed?
[ ] yes [x] no
Does this change modify the behaviour of other functions?
[ ] yes [x] no
Are there code practices which require justification?
[ ] yes [ ] no

### CODE REVIEW

**Comment by coquelin77** [comment]
why is the coverage so low? is it possible to increase this?

**Comment by mtar** [comment]
I added a "header" in the test files to select between CPU and GPU looking for the Environment variable DEVICE. On Travis, only the CPU is used and the GPU part is not covered.

File: heat/core/communication.py [hunk 1]
```diff
@@ -570,32 +604,33 @@ def __allgather_like(self, func, sendbuf, recvbuf, axis, **kwargs):
...
         # perform the scatter operation
         exit_code = func(mpi_sendbuf, mpi_recvbuf, **kwargs)
 
         # undo the recvbuf permutation and assign the temporary buffer to the original recvbuf
-        if axis != 0:
-            recvbuf = recvbuf.permute(*recv_axis_permutation)
-            original_recvbuf.set_(
-                recvbuf.storage(), recvbuf.storage_offset(), recvbuf.shape, recvbuf.stride()
-            )
+        # if axis != 0:
```
Comment (coquelin77) [inline heat/core/communication.py]: remove dead code

File: heat/core/communication.py [hunk 2]
```diff
@@ -713,24 +768,32 @@ def __alltoall_like(self, func, sendbuf, recvbuf, send_axis, recv_axis, **kwargs
...
 
             # perform the scatter operation
             exit_code = func(mpi_sendbuf, mpi_recvbuf, **kwargs)
             # undo the recvbuf permutation and assign the temporary buffer to the original recvbuf
-            if recv_axis != 0:
-                recvbuf = recvbuf.permute(*recv_axis_permutation)
-                original_recvbuf.set_(
-                    recvbuf.storage(), recvbuf.storage_offset(), recvbuf.shape, recvbuf.stride()
-                )
+            # if recv_axis != 0:
```
Comment (coquelin77) [inline heat/core/communication.py]: dead code

File: heat/core/communication.py [hunk 3]
```diff
@@ -953,6 +1079,57 @@ def __getattr__(self, name):
...
+        if (
+            self.tensor is not None
+            and isinstance(self.tensor, torch.Tensor)
+            and self.tensor.is_cuda
+            and not CUDA_AWARE_MPI
+        ):
+            if self.permutation is not None:
+                self.recvbuf = self.recvbuf.permute(self.permutation)
+            self.tensor.copy_(self.recvbuf)
+        return
```
Comment (coquelin77) [inline heat/core/communication.py]: unnecessary, optional remove

File: heat/core/dndarray.py [hunk 4]
```diff
@@ -51,6 +51,19 @@ def __init__(self, array, gshape, dtype, split, device, comm):
         self.__device = device
         self.__comm = comm
 
+        # handle inconsistencies between torch and heat devices
```
Comment (coquelin77) [inline heat/core/dndarray.py]: Can you add this to the `factories.array` function as well?

File: heat/core/dndarray.py [hunk 5]
```diff
@@ -1414,7 +1427,10 @@ def __getitem__(self, key):
                 else:
                     gout = tuple(self.__array[key].shape)
                     if self.split is not None and self.split >= len(gout):
-                        new_split = len(gout) - 1 if len(gout) - 1 > 0 else 0
+                        if len(gout) > 0:
```
Comment (coquelin77) [inline heat/core/dndarray.py]: `new_split = len(gout) - 1 if len(gout) - 1 >= 0 else None`

File: heat/core/logical.py [hunk 6]
```diff
@@ -106,7 +106,10 @@ def allclose(x, y, rtol=1e-05, atol=1e-08, equal_nan=False):
     """
     if np.isscalar(x):
         try:
-            x = factories.array([float(x)])
+            if isinstance(y, dndarray.DNDarray):
+                x = factories.array([float(x)], device=y.device)
+            else:
+                x = factories.array([float(x)])
```
Comment (coquelin77) [inline heat/core/logical.py]: this should still get a device. maybe make a device parameter above it to set the devices of both `x` and `y` to be the same

File: heat/core/logical.py [hunk 7]
```diff
@@ -117,7 +120,10 @@ def allclose(x, y, rtol=1e-05, atol=1e-08, equal_nan=False):
 
     if np.isscalar(y):
         try:
-            y = factories.array([float(y)])
+            if isinstance(x, dndarray.DNDarray):
+                y = factories.array([float(y)], device=x.device)
+            else:
+                y = factories.array([float(y)])
```
Comment (coquelin77) [inline heat/core/logical.py]: see above for x

File: heat/core/manipulations.py [hunk 8]
```diff
@@ -830,10 +842,10 @@ def sort(a, axis=None, descending=False, out=None):
         for idx in np.ndindex(local_sorted.shape[1:]):
             idx_slice = [slice(None)] + [slice(ind, ind + 1) for ind in idx]
 
-            send_count = send_vec[idx][rank]
+            send_count = send_vec.cpu()[idx][rank]
```
Comment (coquelin77) [inline heat/core/manipulations.py]: does this _need_ to be cpu? or can it get any device?

File: heat/core/communication.py [hunk 9]
```diff
@@ -371,55 +372,62 @@ def Recv(self, buf, source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG, status=None):
...
         if isinstance(buf, dndarray.DNDarray):
             buf = buf._DNDarray__array
         if not isinstance(buf, torch.Tensor):
-            return func(buf, dest, tag)
+            return func(buf, dest, tag), None
 
-        return func(self.as_buffer(buf), dest, tag)
+        # in case of GPUs, the memory has to be copied to host memory if CUDA-aware MPI is not supported
+        sbuf = buf if CUDA_AWARE_MPI else buf.cpu()
+        return func(self.as_buffer(sbuf), dest, tag), sbuf
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/communication.py]: Thats's right. For non-blocking communication, I want to keep the sendbuffer in my MPIRequest Wrapper class such that the pointer to the buffer is still valid after leaving the send_like function.

File: heat/core/manipulations.py (hunk 8 above)
Comment (mtar) [reply heat/core/manipulations.py]: I don't know why it's there. send_vec is already on the cpu. cpu() can be removed

File: heat/core/manipulations.py [hunk 10]
```diff
@@ -830,10 +842,10 @@ def sort(a, axis=None, descending=False, out=None):
...
             send_disp = [0] + list(np.cumsum(send_count[:-1]))
 
-            recv_count = send_vec[idx][:, rank]
+            recv_count = send_vec.cpu()[idx][:, rank]
```
Comment (mtar) [inline heat/core/manipulations.py]: Same as above

File: heat/core/logical.py (hunk 6 above)
Comment (mtar) [reply heat/core/logical.py]: If no device is given (i.e. device=None), they have the same default device. In this case, I wanted to change the device implicitly, if a scalar is compared with a tensor.

File: heat/core/dndarray.py (hunk 4 above)
Comment (coquelin77) [reply heat/core/dndarray.py]: this logic is already there, my bad

**Comment by coquelin77** [comment]
The coverage of this branch is low because of the code which is added is tested on GPUs, something that travis does not do.

**Comment by coquelin77** [comment]
The coverage of this branch is low because of the code which is added is tested on GPUs, something that travis does not do. after the manual check, the diff coverage is fine.

**Pull request merged**